core/market/decentralized/src/protocol/discovery.rs

```diff
@@ -34,21 +46,182 @@ pub enum DiscoveryInitError {
 
 /// Responsible for communication with markets on other nodes
 /// during discovery phase.
-#[async_trait]
-pub trait Discovery: Send + Sync {
-    async fn bind_gsb(&self, prefix: String) -> Result<(), DiscoveryInitError>;
+#[derive(Clone)]
+pub struct Discovery {
+    inner: Arc<DiscoveryImpl>,
+}
+
+pub struct DiscoveryImpl {
+    offer_received: HandlerSlot<OfferReceived>,
+    retrieve_offers: HandlerSlot<RetrieveOffers>,
+}
+
+impl Discovery {
+    fn new(mut builder: DiscoveryBuilder) -> Result<Discovery, DiscoveryInitError> {
+        let offer_received = builder.offer_received_handler()?;
+        let retrieve_offers = builder.retrieve_offers_handler()?;
+
+        let inner = Arc::new(DiscoveryImpl {
+            offer_received,
+            retrieve_offers,
+        });
+        Ok(Discovery { inner })
+    }
 
     /// Broadcasts offer to other nodes in network. Connected nodes will
     /// get call to function bound in DiscoveryBuilder::bind_offer_received.
-    async fn broadcast_offer(&self, offer: Offer) -> Result<(), DiscoveryError>;
-    async fn retrieve_offers(&self) -> Result<Vec<Offer>, DiscoveryError>;
+    pub async fn broadcast_offer(&self, offer: Offer) -> Result<(), DiscoveryError> {
+        broadcast_offer(self.inner.clone(), offer).await
+    }
+
+    pub async fn retrieve_offers(&self) -> Result<Vec<Offer>, DiscoveryError> {
+        retrieve_offers(self.inner.clone()).await
+    }
+
+    pub async fn bind_gsb(
+        &self,
+        public_prefix: String,
+        private: String,
+    ) -> Result<(), DiscoveryInitError> {
+        let myself = self.inner.clone();
+
+        log::debug!("Creating broadcast topic {}.", OfferReceived::TOPIC);
+
+        let offer_broadcast_address = format!("{}/{}", private, OfferReceived::TOPIC);
```

Comment:
Inside `into_subscribe_msg` is used for different purpose.
offer_broadcast_address is private address only for market usage, where we will receive messages from broadcast. Every service can register many endpoints that will receive messages from broadcast.

Reply:
I'm not sure if I understand it correctly...
Aren't we supposed to use `<OfferReceived as RpcMessage>::ID` as this private address?

Reply:
AAa ok, now I understand the problem.
So the thing is, OfferReceived isn't directly bound to GSB, so it don't have to be RpcMessage at all. But while implementing Callback, I wanted to reuse some stuff and I took RpcMessage to be parameter to callbacks, because I didn't want to reinvent the wheel.
As I see now, it is confusing. `<OfferReceived as RpcMessage>::ID` isn't used at all, but it is required by RpcMessage.
I could change this in future PRs, but I must check how much RpcMessage benefits to Callbacks code.

Reply:
moved to #326